### DISCUSSION TITLE
How do you change the current language ?

### DISCUSSION THREAD

**Discussion by tmarnet** [sergiodxa/remix-i18next#28] [Q&A] [answered]
Hi 👋 

I can't seem to figure out how to change the current language, and there's no example of it in the README

How would you handle that ?

Thanks !

**Comment by sergiodxa** [comment, marked as answer]
The library detects the language using the following things (in the default order):

- The `log` search param
- A cookie (if you pass one)
- The session (if you pass the sessionStorage)
- The Accept-Language header
- The fallback language you configured.

So if you set `?lng=es` in the URL, and Spanish is in your list of supported languages, then the library will return that in the getLocale method.

If you don't have that, and if configured what cookie to use, you can set the locale in a cookie and the library will use that locale (as far as it's in your list of supported languages).

If you didn't configured a cookie, or if the cookie is not set in the browser, it will try to read from the session (if configured), there it will use (by default) the `lng` key with `session.get("lng")`, but you can change that and store the user locale somewhere else.

Finally, it will try to get the language from the Accept-Language header, this has the list of languages in order of preference the user has configured in his browser or computer in general, if one of those is in your list of supported languages the `getLocale` will return that.

If all of the above didn't work because the search params, or the cookie or the session or the header were not defined or if the language it found in all of them is not supported, then it will return the fallback language you configured.

---

When creating a new RemixI18Next instance, you can pass an `order` option where you can customize the order all of that follows, so you could give preference to the session over the cookie and those over the search param, or even remove one of them and don't use it (e.g. never read from the Accept-Language header).

https://github.com/sergiodxa/remix-i18next/blob/8ab6361122147ea0d63c88ef4fb98295bd8473a9/src/i18next.ts#L52-L61

And when calling `getTranslations`, you can pass a string with the locale instead of the request object, so you could force a locale there if you want to read from somewhere else.

https://github.com/sergiodxa/remix-i18next/blob/8ab6361122147ea0d63c88ef4fb98295bd8473a9/src/i18next.ts#L74-L77

**Comment by adrai** [comment]
@tmarnet I did it like that: https://github.com/sergiodxa/remix-i18next/issues/44#issuecomment-1046791769
